source/fem-dem/cfd_dem_coupling.cc

```diff
@@ -413,15 +701,87 @@ CFDDEMSolver<dim>::check_contact_search_step_dynamic(const unsigned int &)
   return contact_detection_step;
 }
 
+template <int dim>
+inline bool
+CFDDEMSolver<dim>::no_load_balance(const unsigned int &)
+{
+  return false;
+}
+
+template <int dim>
+inline bool
+CFDDEMSolver<dim>::check_load_balance_once(const unsigned int &)
+{
+  bool load_balance_step = (this->simulation_control->get_step_number() ==
+                            this->cfd_dem_simulation_parameters.dem_parameters
+                              .model_parameters.load_balance_step);
+
+  if (load_balance_step || checkpoint_step)
+    {
+      load_balance();
+    }
+
+  return load_balance_step;
+}
+
+template <int dim>
+inline bool
+CFDDEMSolver<dim>::check_load_balance_frequent(const unsigned int &)
+{
+  bool load_balance_step = (this->simulation_control->get_step_number() %
+                              this->cfd_dem_simulation_parameters.dem_parameters
+                                .model_parameters.load_balance_frequency ==
+                            0);
+
+  if (load_balance_step || checkpoint_step)
+    {
+      load_balance();
+    }
+
+  return load_balance_step;
+}
+
+template <int dim>
+inline bool
+CFDDEMSolver<dim>::check_load_balance_dynamic(const unsigned int &)
+{
+  bool load_balance_step = false;
+  if (this->simulation_control->get_step_number() %
+        this->cfd_dem_simulation_parameters.dem_parameters.model_parameters
+          .dynamic_load_balance_check_frequency ==
+      0)
+    {
+      unsigned int maximum_particle_number_on_proc = 0;
+      unsigned int minimum_particle_number_on_proc = 0;
+
+      maximum_particle_number_on_proc =
+        Utilities::MPI::max(this->particle_handler.n_locally_owned_particles(),
+                            this->mpi_communicator);
+      minimum_particle_number_on_proc =
+        Utilities::MPI::min(this->particle_handler.n_locally_owned_particles(),
+                            this->mpi_communicator);
+
+      if ((maximum_particle_number_on_proc - minimum_particle_number_on_proc) >
+            this->cfd_dem_simulation_parameters.dem_parameters.model_parameters
+                .load_balance_threshold *
+              (this->particle_handler.n_global_particles() /
+               this->n_mpi_processes) ||
+          checkpoint_step)
+        {
+          load_balance();
+          load_balance_step = true;
+        }
+    }
+
+  return load_balance_step;
+}
```

Comment:
Again, the logic of this function could be refactored out of the CFD-DEM and DEM classes and the boolean value could then just be used.

Reply:
done